[Blackjack - Step2] Submit san-ghun's mission

san-ghun opened the pull request:
# Blackjack Step 2

## Overview

This PR is to implement requirements of the Blackjack project step2. The project is a simulation of card game Blackjack in terminal text input and output. The step2 requirement is to add betting feature in the version of Blackjack. In this game, player(s) bet certain amount, the player or dealer with a total closest to 21 without going over wins, and finally display earnings for player and dealer based on player's betting and game result.

## Main Changes

### `Result` has new enum `BLACKJACK` 

- It is to mark the result of player or dealer who has blackjack, for player to receive 1.5x of their earning.

### `Player` can bet before the initial draw of cards

- `Playable` has a property named `bet` with integer value to store betting amount.
- `Playable` has a method `isBlackjack()` to figure out the `Playable` is blackjack or not.
- In 'InputView', `readPlayerBettingAmount()` read user input of player's betting amount.
- `Player` has a method `placeBets()` to add betting amount to on its `bet` property.

### Output game result with earnings of all `Playable`s

- `Stats` calculate pay-outs for players and dealer based on the conditions;
    - All betting values are added on a variable `pot` at during the `payOutPotToEarnings()` method.
    - Check the dealer has blackjack or not.
    - Check the player has blackjack or not.
    - All players get pay-outs from `pot`, then dealer get reminder from `pot`.
- 'OutputView' display the final result and earnings with `displayEarnings()`.

## Additional Notes

- Try to simplify the interaction between functions to improve readability.
- Try to remove states and replace those into methods in class.

In this project, coding logic for calculating the earnings for `Playable`s was difficult, especially to separate the logic into several functions to meet the Programming Requirements of this project. Also tried to apply reviewed comments from previous step1, for example, replacing states into methods.

Thank you!

san-ghun committed:
docs(readme): update README.md file for beginning step2 mission

san-ghun committed:
feat(InputView): add method 'readPlayerBettingAmount()' to read user input for player betting

san-ghun committed:
feat(Playable): add property 'bet' and method 'isBlackjack()'

san-ghun committed:
feat(Player): add method 'placeBets()'

san-ghun committed:
refactor(Player,Dealer,GameMaster): modify 'Player' and 'Dealer' internally init 'hand' and simplify initiating 'players'

san-ghun committed:
refactor(result): add new enum 'BLACKJACK' to give result of blackjack to 'Playable's

san-ghun committed:
refactor(Player,Dealer,GameMaster): add method 'initHand()' for 'Player' and 'Dealer' to init hand with two cards

san-ghun committed:
feat(Stats): remove 'updateDealerStats()' and add method 'payOutPotToEarnings()' to calculate earnings for players and dealer

san-ghun committed:
feat(OutputView,GameMaster): add 'displayEarnings()' to display the final earnings for players dealer

san-ghun committed:
fix(Stats): add blackjack check for player to have result of blackjack

Wordbe commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -14,6 +15,11 @@ data class Dealer(override val name: String = "Dealer") : Playable {
+    fun initHand(): Dealer {
+        hand = Hand().initCards()
+        return this
+    }
```
If you store the hand directly during Dealer initialization, I don’t think `initHand()` is necessary. Just calling the constructor should be enough. What do you think? ☺️

san-ghun replied:
Yeah, your right. 
I was wondering how I could write tests that could inject a set of cards and control conditions to keep the outcome of the `Playable` predictable, without having to control the injection of cards into the `Playable` using `initHand()`.

I come up with this solutions, :

1. Let's inject `Hand` into `Playable`.

```kotlin
data class Dealer(
    override val name: String = "Dealer",
    override val hand: Hand = Hand(),
) : Playable {
//...
}
```

2. Let `Hand` to init with condition.

```kotlin
class Hand(var cards: List<PlayingCard> = emptyList()) {
    val isBust: Boolean get() = calculateHand() > BUST_LIMIT
    val isBlackjack: Boolean get() =
        (cards.size == INIT_CARDS_SIZE) && (calculateHand() == BUST_LIMIT)

    init {
        if (cards.isEmpty()) {
            val deque = ArrayDeque(cards)
            repeat(INIT_CARDS_SIZE) {
                deque.addLast(PlayingCard.deck.giveCard())
            }
            cards = deque.toList()
        }
    }
//...
}
```

Wordbe commented on src/main/kotlin/blackjack/controller/GameMaster.kt:
```diff
@@ -11,54 +10,42 @@ import blackjack.view.OutputView
+        val bets = names.map { InputView.retryable { InputView.readPlayerBettingAmount(it) } }
+        val players =
+            names.zip(bets)
+                .map { (name, amount) -> Player(name).placeBets(amount).initHand() }
```
Using `zip` is a great choice! 😊 That said, it might be better to stick to just one term—either `bet` or `amount`—for consistency. What do you think?

san-ghun replied:
I agree, good point! Thank you!

Wordbe commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -17,7 +18,14 @@ interface Playable {
     }
 
+    fun isBlackjack(): Boolean {
+        val hasTwoCards = hand.cards.size == 2
```
I think `hasTwoCards` could be a responsibility of the `Hand` class.
Also,try to create a test of `Hand` and eliminate the magic number 😊

san-ghun replied:
Yes, it seems better fit with `Hand`!

Wordbe commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -17,7 +18,14 @@ interface Playable {
 
+    fun isBlackjack(): Boolean {
+        val hasTwoCards = hand.cards.size == 2
+        val hasTwentyOne = calculateHand() == BUST_LIMIT
```
Now that I look at it, `calculateHand()` could be implemented directly in the interface since both implementations use the same logic 😊
With that in mind, I think `hasTwentyOne` could also be a responsibility of the `Hand` class.

san-ghun replied:
Agree! I move this to `Hand`.

Wordbe commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -12,4 +14,14 @@ data class Player(override val name: String, override var hand: Hand = Hand()) :
         return hand.calculateHand()
     }
+
+    fun initHand(): Player {
+        hand = Hand().initCards()
+        return this
+    }
+
+    fun placeBets(amount: Int): Player {
+        bet += amount
+        return this
+    }
```
Same here as with the Dealer—looks like the constructor can handle it just fine! 😊
Also, try adding a validation check inside the constructor.

san-ghun replied:
Yes, I apply same way to init `Player` with constructor like did on `Dealer`.
However, I didn't get what you mean that, try adding a validation check inside the constructor, for which validation you mean?

Wordbe replied:
I meant it would be good to validate whether `amount` is a positive number! 🙂

Wordbe commented on src/main/kotlin/blackjack/model/Result.kt:
```diff
@@ -4,5 +4,11 @@ enum class Result() {
+    ;
+
+    companion object {
+        const val BLACKJACK_BONUS = 1.5
```
Specifying the type makes it easier for others to read the code 🙂

san-ghun replied:
Yes, it would be better with type notation. I will keep that in mind! Thanks!

Wordbe commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -39,25 +38,14 @@ object OutputView {
+        val players = winStatistics.players
+        val dealer = winStatistics.dealer
+        val earningsMap = winStatistics.payOutPotToEarnings()
+        println("\n## Final Earnings")
+        println("Dealer: ${earningsMap[dealer]}")
```
How about calling `println` once to check? 🙂

san-ghun replied:
What you want to say is, make the two lines of `println` into one line, like below?
So, to reduce the amount of calling `println` function?
Did I understood correct?

```kotlin
println("\n## Final Earnings\nDealer: ${earningsMap[dealer]}")
```

Wordbe replied:
I was referring to using the `"""` syntax for string templates. Please take a look! 📄
https://kotlinlang.org/docs/strings.html#string-templates

Wordbe commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -1,20 +1,6 @@
 package blackjack.model
 
 class Stats(val players: List<Player>, val dealer: Dealer) {
```
The game result calculation and payout logic are mixed together—it would be better to separate them 😊
Also, can you try simplifying the code a bit more?
For example, instead of subtracting each player's `payoutImpactOnPot` from the pot, you could try something like:

```kotlin
val dealerEarnings = totalBets - totalPlayerPayouts
```

This way, you can eliminate mutable variables.
Try making good use of functional programming too!

san-ghun replied:
Yes, I can see some chance to simplify the code. I will aggregate both logic into one and make it simple!

Wordbe replied:
So great. Thank you!

Wordbe reviewed (changes requested):
Hello San!

Thank you for understanding and implementing the additional requirements of the Blackjack mission so well.
I’ve left comments on some points that need further thought, so please review them and apply any changes where possible.
Also, please strengthen the tests for the parts you modified, if you can (Dealer, Playable).

Thank you! 😊

san-ghun committed:
style(GameMaster): make expression more consistent to improve readability

san-ghun committed:
style(Result): rename with type notation to improve readability

san-ghun committed:
style(OutputView): consolidate two lines of 'println' into one

san-ghun committed:
refactor(Playable,Hand): let 'Playable' init its property with constructor and let hand to init with condition

- move some of unnecessary method with responsibility from 'Playable' to 'Hand'

san-ghun committed:
refactor(Stats): aggregate and simplify payout and earning calculation logic

san-ghun commented:
Hello Suno!

Wish you the best of this summer.
Always thanks for your review with valuable suggestions.
I tried to apply your reviews.

## Main Changes

### Move actions and properties from `Playable` to `Hand`

- transfer calculate-able properties `isBust` and `isBlackjack` to check the status of 'hand' from `Playable` to `Hand`.
- remove `calculateHand()` from `Playable` and call to `Hand` directly to calculate the hand.
- remove `initHand()` from `Player` and `Dealer` to `Hand` by applying `init{}` with condition to initiate the `Hand`.
- remove `placeBets()` from `Player` and use constructor to get `bet` amount, since the betting is done at once at the beginning of the game.

### Aggregate and simplify calculation of payout and earning of `Playable`s

- aggregate and simplify the calculation of payout and earning of `Player` and `Dealer` by applying the functional programming approach.

## Additional Note

Through your comments and review, I tried to apply functional programming as much as possible. I realized that there are plenty way of using the functional programming approach to implement even a single logic. It was fascinating to have multiple choices and to find out more options by chaining other methods.

Also, I found that it is essential to have an Object-Oriented mindset like using constructor efficiently and take advantage of OOP way. I am still finding out and learning the OOP way through projects and also from reading articles about the Design Patterns. Those are still hard to understand and apply for me, however, I am getting more curious what will come next and how my code will become more magical like Jason, Brie and Justin, our coaches.

I really appreciate your patience for my request for review.
It took long time to submit this request for review and I apologize for taking this long time.
However, thanks to you patience I could have more time to consider how to use functional programming to to simplify complex logic and apply the constructor to initialize an instance in more OOP way.
Thanks for keep me motivated through your thoughtful review and comments.

Wordbe reviewed (approved):
Hello San!

You didn’t give up and completed the mission to the end. great job!
Thank you for carefully reading through the reviews and applying the feedback.

I hope what you’ve learned from this mission will be useful in future projects as well.
If you take some time to revisit the following keywords, it’ll help solidify your understanding:

* Proper use of inheritance and interfaces
* Assigning responsibilities to objects to increase cohesion
* Applying polymorphism

I'll go ahead and approve and merge this PR now.
It would be great if you could also reflect on the additional comments I left.

Thank you again for your dedication all the way through! 🙌
